src/server/raft_client/fresh.rs

```diff
@@ -610,25 +658,216 @@ where
             Err(e) => {
                 RESOLVE_STORE_COUNTER.with_label_values(&["failed"]).inc();
                 back_end.clear_pending_message();
+                // TOMBSTONE
+                if format!("{}", e).contains("has been removed") {
+                    let mut pool = pool.lock().unwrap();
+                    if let Some(s) = pool.connections.remove(&(back_end.store_id, conn_id)) {
+                        s.disconnect();
+                    }
+                    pool.tombstone_stores.insert(back_end.store_id);
+                    return;
+                }
                 continue;
             }
         };
         let client = back_end.connect(&addr);
-        let mut res = back_end.batch_call(&client, addr.clone()).await;
+        let f = back_end.batch_call(&client, addr.clone());
+        let mut res = f.await;
         if res == Ok(()) {
             // If the call is setup successfully, it will never finish. Returning `Ok(())` means the
             // batch_call is not supported, we are probably connect to an old version of TiKV. So we
             // need to fallback to use legacy API.
-            res = back_end.call(&client, addr.clone()).await;
+            let f = back_end.call(&client, addr.clone());
+            res = f.await;
         }
         match res {
             Ok(()) => {
                 error!("connection fail"; "store_id" => back_end.store_id, "addr" => addr, "error" => "require fallback even with legacy API");
             }
             Err(_) => {
                 error!("connection abort"; "store_id" => back_end.store_id, "addr" => addr);
-                back_end.router.broadcast_unreachable(back_end.store_id);
+                back_end
+                    .builder
+                    .router
+                    .broadcast_unreachable(back_end.store_id);
+            }
+        }
+    }
+}
+
+/// A global connection pool.
+///
+/// All valid connections should be stored as a record. Once it's removed
+/// from the struct, all cache clone should also remove it at some time.
+#[derive(Default)]
+struct ConnectionPool {
+    connections: HashMap<(u64, usize), Arc<Queue>>,
+    tombstone_stores: HashSet<u64>,
+}
+
+/// Queue in cache.
+struct CachedQueue {
+    queue: Arc<Queue>,
+    /// If a msg is enqueued, but the queue has not been notified for polling,
+    /// it will be marked to true. And all dirty queues are expected to be
+    /// notified during flushing.
+    dirty: bool,
```

Comment:
why not use `queue.len() > 0`

Reply:
`queue.len() > 0` just means there are data in the queue, it can't be tell whether it has been notified.